packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -104,16 +117,61 @@ contract L2ToL2CrossDomainMessenger is IL2ToL2CrossDomainMessenger, ISemver, Tra
         if (_target == Predeploys.CROSS_L2_INBOX) revert MessageTargetCrossL2Inbox();
         if (_target == Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert MessageTargetL2ToL2CrossDomainMessenger();
 
+        /*
         bytes memory data = abi.encodeCall(
-            L2ToL2CrossDomainMessenger.relayMessage,
+            L2ToL2CrossDomainMessenger.execute,
             (_destination, block.chainid, messageNonce(), msg.sender, _target, _message)
         );
         assembly {
             log0(add(data, 0x20), mload(data))
         }
+         */
+
+        emit SentMessage(_destination, _target, messageNonce(), msg.sender, _message);
+
         msgNonce++;
     }
 
+    function relayMessage2(ICrossL2Inbox.Identifier calldata _id, bytes calldata _sentMessage) external payable nonReentrant {
+        // Ensure the log came from the messenger and validate with the cross l2 inbox
+        if (_id.origin != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert("not the l2 cdm");
+        CrossL2Inbox(Predeploys.CROSS_L2_INBOX).validateMessage(_id, keccak256(_sentMessage));
+
+        // Open Question:
+        //   - should we assert length on `_sentMessage` prior to decoding? It's structure should implicitly be known and valid
+        //   as _id is already validated via the inbox
+        //
+        //   - How does this work between upgrades? The log data wouldn't be evolvable without invalidating unsent messages. New
+        //   entrypoint / relay mechanisms must be introduced? Or the `log.Data` could be parsed after decoding the versioned
+        //   message nonce? However this would break the event signature
+
+        // decode the topics (skipping the event signature. Maybe we want to validate this?)
+        (uint256 _destination, address _target, uint256 _nonce) = abi.decode(_sentMessage[32:128], (uint256,address,uint256));
```

Comment:
Thoughts on doing a single abi decode, including the selector?

```
(
  bytes32 selector
  uint256 _destination,
  address _target,
  uint256 _nonce,
  address sender,
  bytes memory message
) = abi.decode(_sentMessage, (bytes32,uint256,address,uint256,address,bytes));

if (selector != SentMessage.selector) revert InvalidMessage();
```

Reply:
strangely when I try this I get a `panic: memory allocation error` when running the test with verbosity in forge, validating the selector is a good idea.

might be a forge issue, will look into it

Reply:
It could be due to an extra layer of abi encoding